### ISSUE TITLE
[Popup] Should the delegatesfocus attribute be renamed?

### ISSUE THREAD

**Issue by melanierichards** [openui/open-ui#368]
In PR #355, @mfreed7 mentioned on a discussion regarding the `delegatesfocus` attribute, which tells the user agent to move focus to the popup's first focusable descendent when the popup is shown:

> If you're referring to [this version of delegatesfocus](https://dom.spec.whatwg.org/#dom-shadowroot-delegatesfocus), that's fairly different. That refers to Shadow DOM and how focus is proxied into the shadow root (and isn't well specified, at least in DOM/HTML specs). I bring this up for two reasons:
> 
> I do think this version here needs a specific definition like this, to be clear about what it does.
> I am wondering if we should overload this attribute name for two similar but different concepts. Not sure if it helps or hurts developer understandability.

Opening this issue to generate discussion on whether `delegatesfocus` is the right name for this attribute.

**Edit:** when this to-be-renamed attribute is applied to `popup`, focus will be automatically sent to the first tab-focusable descendant of the popup, when that popup is shown. So for example:

```
<popup autofocus delegatesfocus>
  <p>I'm just some friendly text</p>
  <button>I will be focused, not the popup itself</button>
</popup>
```

The `<button>` would get focus when this `<popup>` is shown. Otherwise, focus would remain wherever it was when the popup was shown; alternatively, a dev could add the `autofocus` attribute to `<popup>`, and the `<popup>` itself would take focus.

Note: we are discussing on #367 how this attribute should be extended to other elements on the platform. There's sort of an open question as to whether:

* This would be applicable to other elements that don't have a `show()` event, as `popup` and `dialog` do
* If so, whether it would:
  * Automatically steal focus
  * Only send focus from an element to its descendant when the _element itself_ is meant to receive focus (e.g. user tabs to element with `delegatesfocus` applied)
  * Have some option to specify either of the prior behaviors
  * Be composable with `autofocus` (i.e. `delegatesfocus` does not steal focus unless the element to which it's applied also has `autofocus`) <-- IMHO this could be a nice idea

**Comment by chrisdholt** [comment]
I don't know that I have a better name or semantic here, but I would agree on renaming - I saw this and got confused as to how this was related to delegating focus to a shadow root.

**Comment by bkardell** [comment]
catching up on details now, and I didn't realize at first this was _not_ regular delegates focus because they do both ...delegate focus.  I guess that means something different in both contexts that is unfortunately unclear... I suppose the native one kind of 'proxies' while this one kind of 'redirects' (to use webby terms I think we might have similar understanding of)?

**Comment by gregwhitworth** [comment]
This was discussed in the telecon today and we resolved as follows:

> Rename `delegatesfocus` content attribute. Name TBD

**Comment by melanierichards** [comment]
As part of the resolution, folks please feel free to chime in on this issue with a suggestion for an HTML content attribute with the behavior "send focus to my first tab-focusable descendent when I am shown".

**Comment by melanierichards** [comment]
Trying to generate some ideas for bikeshedding an attribute name for "send focus to my first tab-focusable descendent instead of myself"...

* `transfersfocus`: does that seem like focus could be sent to a non-descendent?
* `bequeathsfocus`: "bequeath" is the opposite of "inherit" but it seems a little fancy 🤵 for an attribute
* `focusdescendant` / `focusdescendants`
* `passesfocus`
* These imply a particular behavior if the attribute is extended to non-event-driven elements, like a `div` instead of a `popup` or `dialog` (i.e. descendants would get focused immediately on "page load"):
  * `autofocuschildren`
  * `autofocusdescendants`
  * `autofocuscontent`
* `passthroughfocus`

**Comment by bkardell** [comment]
I think I offered `relaysfocus` or `redirectsfocus` on the call as options that made sense to me.

**Comment by melanierichards** [comment]
Thanks @bkardell, forgot about those. `relaysfocus` seems short and sweet and clear.

**Comment by Wolfr** [comment]
'passfocus' seems like a winner to me.

**Comment by maxsteenbergen** [comment]
How about `entrustFocus`?

**Comment by bkardell** [comment]
`routesfocus` came to me later, I guess route and relay seem like good webby terms people can kind of relate to and might mentally conjure up the relationships - is my thinking

**Comment by hidde** [comment]
+1 to Melanie's `transfersfocus` and `passesfocus`, and Brian's suggestions of `relaysfocus` and `redirectsfocus`.

**Comment by diondiondion** [comment]
>       * `autofocuschildren`
>       * `autofocusdescendants`
>       * `autofocuscontent`

I like these over the other suggestions. IMO the ideal "universal" API would be something like `autofocus="within"`. But I guess re-using an existing attribute is off the table.

I was also thinking about why `autofocus` works so well as an attribute name: it treats "focus" as a verb. Most other suggestions combine it with another verb, but to me the added verbs don't add much meaning and all feel more or less interchangeable (might as well go fully generic: `handlefocus` or `setfocus`). In the end I'm always left wondering _what_ will be focused.

So I'd keep looking more in the direction of something like `focuscontent` or `focuswithin`.

**Comment by bradkemper** [comment]
Focus-within

**Comment by hidde** [comment]
`focus-within` would be too confusing I think, because in CSS the pseudo class with that name reflects some sort of state (does an element within this element have focus or not), not an action, and it is true if any child has focus, while delegatesfocus would delegate to one specific child: the first focusable one.

**Comment by melanierichards** [comment]
@diondiondion said:

> IMO the ideal "universal" API would be something like autofocus="within". But I guess re-using an existing attribute is off the table.

I want to call out that yours is an interesting idea, and maybe `autofocus` is a boolean that could've/should've been an enum. :) If it were just about the user agent handling the attribute value appropriately, there's probably some logic that could work there even if it's a little messy (treat `autofocus` like `autofocus="self"`, recognize `autofocus="within"`, throw away any other keywords). Probably where we'd be more likely to run into compat issues would be getting/setting the IDL attribute, and any scripts that assume a boolean value. So, I think we would need to introduce something different but just wanted to say thanks for exploring this concept!

**Comment by melanierichards** [comment]
## Putting this to a vote

Now that we have a good collection of options, we'd like to put this attribute to a vote. I've collected a couple options that have multiple people interested. Please use the emoji reaction function on this comment to vote on one of the following options:

👍 = `relaysfocus` (can change tense to `relayfocus`)
🎉 = `redirectsfocus` (or `redirectfocus`)
♥ = `transfersfocus` (or `transferfocus`)
🚀 = `passesfocus` (or `passfocus`)
👀 = `focuscontent`

If for some reason you run into issues using the reaction function to vote, or don't have a Github account, please reach out and we'll find a way to make sure your input is counted. e.g., I will probably share this on Twitter, feel free to express your opinion as a direct response to the tweet.

**Comment by BrandonLive** [comment]
As a data point, I came here and saw the options before I knew what the proposed attribute actually will do, and I would not have intuited that behavior from any of these options. The last one, “focuscontent” came the closest for me. The others seem like they’d be better fit for a scenario where you wanted to actually forward focus from one element to another (e.g. when clicking this button, I want to focus a particular text box).

It seems this attribute will enable a side effect whenever the popup is shown, which led me to think of somewhat verbose names like “focuscontentonshow”.

It also seems this could also be viewed as controlling two behaviors:
- Taking focus when shown
- Routing/forwarding focus to the first focusable descendant in the tab order

I wonder if there’s merit to teasing those apart?

The first of those almost seems like it could be a parameter on the show() method, rather than an attribute. But perhaps there are other ways these get shown? Or just reuse autofocus for that?

The ideas about extending the “autofocus” attribute with multiple possible values are compelling, it’s a shame that compatibility concerns apparently make that approach unviable.

**Comment by JamesCoyle** [comment]
I agree with @BrandonLive here. All but `focuscontent` sound like they transfer focus on a particular user action rather than automatically focusing on element. 

I think `autofocus="content"` would be ideal here or `autofocuscontent` if there are issues adding functionality to the existing autofocus attribute.

**Comment by diondiondion** [comment]
> So, I think we would need to introduce something different but just wanted to say thanks for exploring this concept!

Thanks for the response @melanierichards.

Since autofocus can't be changed, how about introducing a new enum attribute `initialfocus` instead of yet another boolean? E.g., `initialfocus="self"` could be equivalent to autofocus, `initialfocus="within"` (or similar) could be used for the new behaviour. This way at least the new attribute will be open to future extension.

**Comment by getflourish** [comment]
Throwing in the concept of **responders** from Apple UIKit (https://developer.apple.com/documentation/uikit/uiresponder)

For HTML, I’ve leaned on that concept, often calling a method "focusFirstResponder" when I open a dialog or similar things.

If re-using autofocus is not an option, I can imagine a more generic attribute such as `focus` with a clear value, such as:

- `focus="firstResponder"` 

Or in the context of HTML:
 
- `focus="content"`
 
The question would be wether there are other use cases other than focussing the first responder? Depending on that, it could be boolean or rather have a set of options.

I could imagine `focus="skip"` or `focus="off"` as a more explicit way compared to specifying `tabindex="-1"`.

**Comment by no-identd** [comment]
This whole discussion reminds me of the [Microsoft Powershell approved verbs list](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.2).

**Comment by giuseppeg** [comment]
Why not allow `autofocus` only on one element? Either the popup or button? After all you can't focus both?

Case 1. autofocus is on the popup and the popup is focused

```html
<popup autofocus>
  <button>ok</button>
</popup>
```

Case 2. autofocus is on the button and the button is focused

```html
<popup>
  <button autofocus>ok</button>
</popup>
```

Case 4. autofocus is not set and nothing gets focus automatically

```html
<popup>
  <button>ok</button>
</popup>
```

**Comment by mfreed7** [comment]
Reading back through this issue, it sounds like `focuscontent` seems to be the winner. It got the most [votes](https://github.com/openui/open-ui/issues/368#issuecomment-892244537), and has a number of supporting comments.

Can we resolve to rename to `focuscontent`?

**Comment by mfreed7** [comment]
I'm going to Agenda+ this to discuss at a future meeting. There seem to be a number of options:
- Go with `focuscontent` as it [won the vote](https://github.com/openui/open-ui/issues/368#issuecomment-892244537).
- Add some late-breaking suggestions like `autofocuscontent` or `focus=foo` and vote again.
- Remove `delegatesfocus` from the MVP for the Popup API, and keep discussing.

I'm beginning to be inclined to vote for the last option. Especially since we want to [expand this attribute to other elements](https://github.com/openui/open-ui/issues/367), there are [issues with doing that](https://github.com/openui/open-ui/issues/379), and the popup-specific use cases are [nebulous](https://open-ui.org/components/popup.research.explainer#light-dismiss:~:text=The%20purpose%20of%20this%20attribute%20is%20to%20handle%20the%20cases) at best. If a `delegatesfocus`-like attribute is added generally to the platform, it'll benefit Popups along with everything else.

So basically, Agenda+ to ask for a resolution removing `delegatesfocus` from the MVP for Popup.

**Comment by scottaohara** [comment]
the use case for `delegatefocus` is... if you have a popup where you know you need to set focus to something within it, but you don't know what the content of the popup will be, and hence you can't just use `autofocus` on the element which should receive focus?

**Comment by mfreed7** [comment]
> the use case for `delegatefocus` is... if you have a popup where you know you need to set focus to something within it, but you don't know what the content of the popup will be, and hence you can't just use `autofocus` on the element which should receive focus?

Right, but I guess I'm wondering if that use case actually comes up, and in what context (and how often)?
